### ISSUE TITLE
Stop Loss e Timelapse (Perdas e Tempo de trade excessivos em alguns casos)

### ISSUE THREAD

**Issue by moserbtc** [moserbtc/Freqtrade_Confluence_Strategy#1]
Vamos analisar em detalhes por que o backtesting continua exibindo (1) durações excessivas em alguns trades, (2) perda muito grande em determinados casos. O código em si parece correto na lógica de time-based stop e no cálculo de TP/SL, mas há diversas razões pelas quais isso pode não mudar o comportamento no backtest. Abaixo estão as principais frentes de investigação:

1. Checar se use_custom_stoploss realmente está habilitado

Verifique se, no seu config.json, há algo assim:

{
  // ...
  "use_custom_stoploss": true,  // <= Precisamos disso ligado!
  // ...
}

Em versões recentes do Freqtrade, não é mais preciso declarar explicitamente use_custom_stoploss: true; basta que sua estratégia implemente custom_stoploss e não tenha trailing_stop habilitado. Entretanto, em algumas versões, se use_custom_stoploss estiver false ou ausente, a engine pode simplesmente ignorar sua função custom_stoploss.

Como testar
	•	Rode o freqtrade backtesting com debug maior (-v 3 ou --logfile) e procure nos logs algo como Custom stoploss disabled. Se isso estiver aparecendo, significa que a engine não chama a sua função custom_stoploss.

2. Garantir que max_trade_duration_in_hours não seja muito grande

Você mencionou ver trades durando 20, 50, 600 dias. Se max_trade_duration_in_hours=72, teoricamente nenhum trade deveria ultrapassar 3 dias. Se no relatório final ainda há trades com 100+ dias, só há duas explicações:
	1.	O time-based stop não é chamado: Falha na habilitação do custom stoploss (ponto anterior).
	2.	Há lógica que substitui ou ignora esse custom stop: Por exemplo, liquidação, force_exit no final do timerange, etc.

Ver também

Se você executa o backtest com um período muito extenso (por exemplo, 2022 a 2025) e vê trades com “277 days”, pode ser que o trade tenha começado lá no início e nunca atingiu o SL/TP, só terminando no final do backtesting – caso o time-based stop não esteja de fato desligando a posição.

3. Ver se a “liquidação” em Futures está ativada

Quando se opera --trading-mode futures, a engine do Freqtrade faz simulação de liquidação se a perda alcança determinado nível. Isso, às vezes, gera perdas enormes (-80% etc.) se a alavancagem simulada estiver alta.
Se você não quer ver liquidação, ou quer permitir margin calls diferentes, pode ser preciso desativar ou configurar a simulação de liquidação:
	•	No config.json, remover ou ajustar:

"position_adjustment_enable": false,
"unrealized_profit_threshold": ...

	•	Ou usar menor alavancagem se esse for um caso.

4. Checar force_exit no final do timerange

Muitas vezes, trades que ficam abertos até o fim do período de backtest são marcados como force_exit. Se o backtest pegou todo o range 2022-01 até 2025-01, e seu time-based stop não está efetivamente fechado o trade, ele vai sair forçado no final. Se no relatório aparece Avg. Duration Loser: 50+ days, é o trade que abriu “lá atrás” e só fechou force_exit no final do timerange.

Como confirmar

Veja no EXIT REASON STATS a quantidade de force_exit. Se trades losers aparecendo com force_exit e grande duração, você já sabe que não passou pelo custom_stoploss.

5. Confirmar se a "stoploss": -0.99 não está sendo acionada antes do custom_stoploss

Com “stoploss = -0.99”, a engine do Freqtrade nativo tem um “stop” de 99% de perda, mas como você está usando Futures, pode haver multiplicações de alavancagem ou a perda ser calculada de modo diferente. Mesmo assim, se custom_stoploss está habilitado, ele deveria tomar precedência sobre esse valor estático.
Caso esse “-0.99” esteja conflitando com liquidação ou algo assim, pode permitir trades passarem de -80% etc. Pois “-0.99” é de fato 99% de perda, muito distante.

6. Conflito com trailing_stop

Se, no config.json, houver algo como

"trailing_stop": true

isso pode “desligar” ou conflitar com a forma como o custom_stoploss funciona, dependendo da versão do Freqtrade. Confira se você realmente desativou o trailing stop no config e na estratégia.
O seu código mostra trailing_stop = False, mas se o config for true, pode acontecer uma confusão.

7. Adicionar logs dentro do custom_stoploss

Para ter certeza de que o time-based stop está sendo chamado, adicione logs (ex: print ou logger.info) dentro da função custom_stoploss. Por exemplo:

def custom_stoploss(...):
    print(f"[DEBUG STOPLOSS] trade.id={trade.id}, side={side_}, open_date={trade.open_date_utc}, current_time={current_time}, dur_hours=XXX")
    # ...

Rode o backtest com --logfile=debug.log. Se não aparecer nada desses logs, é prova de que o método não está sendo chamado.

8. Testar com período menor ou com config minimal

Para isolar se o problema é do time-based stop, faça um backtest rápido de 1 mês:

freqtrade backtesting \
  --strategy IntegrativeConfluenceStrategy \
  --timerange=20230101-20230201 \
  --config user_data/config.json \
  --logfile debug.log

E veja se ainda ocorrem trades de 300 dias. Se sim, é 100% prova de que o time-based stop está ignorado.

9. Parametrizar um Stop min_stop_dist_factor maior

Em alguns mercados, se o ATR estiver muito pequeno, seu STOP pode acabar calculado tão perto que rejeita o trade. Ou se ATR estiver grande, seu STOP acaba enorme. Você já tem um range 0.3 ... 1.0 para min_stop_factor.
Tente aumentar min_stop_factor para ~1.0 e baixar max_stop_factor para ~2.0. Assim, stops ficam mais curtos, e a chance de durar ~600 dias tende a zero.

10. Confirmar se calc_takeprofit_stoploss_v4 gera o STOP do jeito esperado

Você pode colocar debug=True dentro do calc_takeprofit_stoploss_v4, momentaneamente, e ver prints dos valores de dist_tp, stop_dist etc. Se os logs mostram STOP gigante (ex.: 1000 USDT de dist no BTC, que seria uma variação imensa), então vai demorar pra bater.

11. Conclusão:

Se as mudanças no código não alteraram o backtest, é quase certo que:
	1.	O custom_stoploss não está sendo invocado (por config/versão) ou
	2.	Há liquidação/force_exit que ocorre e time-based stop não prevalece.

Geralmente, o passo essencial é conferir nos logs se, de fato, o custom_stoploss imprime algo (ou o debug). Se não imprime, o Freqtrade não está chamando seu método. Nesse caso, ajuste no config.json ou reavalie a versão do Freqtrade.
—————————
Depois que o custom_stoploss estiver realmente ativo, e com max_trade_duration_in_hours=72, você não deve mais ver trades se estendendo por dezenas de dias. E trades com -80% de perda também tenderão a sumir (ou encerrar-se antes) – a não ser que haja liquidação simulada em Futures com alavancagem alta.

A discrepância entre a duração média dos trades vencedores e a duração média dos trades perdedores (por exemplo, Winners: ~4 dias vs. Losers: ~56 dias) costuma indicar que alguns trades “perdedores” estão ficando abertos por muito tempo, o que normalmente não deveria ocorrer se o time-based stop estiver configurado corretamente. Abaixo, listo possíveis causas e formas de investigar por que o time-based stop ou o stoploss podem não estar sendo acionados adequadamente em certos casos:

1. use_custom_stoploss desativado
	•	Sintoma: Mesmo que exista custom_stoploss na estratégia, se o config.json não tiver use_custom_stoploss: true (ou equivalente), o Freqtrade não chamará sua função custom_stoploss no backtest.
	•	Conferir:
	•	Se no seu config.json existe algo como:

"use_custom_stoploss": true

ou se, dependendo da versão, a custom_stoploss não foi desativada por outro parâmetro.

2. max_trade_duration_in_hours muito alto
	•	Sintoma: Se a duração máxima configurada for, por exemplo, 240 horas (10 dias) ou um valor ainda maior, trades podem ficar abertos por muito mais tempo que o desejado.
	•	Exemplo: Se setado em 72 horas, esperado é que ninguém fique além de ~3 dias. Se você vê 56 dias, é provável que não esteja sendo respeitado.
	•	Conferir:
	•	Se a max_trade_duration_in_hours está, de fato, em 72 (ou outro valor pequeno) e se o backtest não estiver usando esse valor.

3. Possível falha ao parsear current_time
	•	Sintoma: Pode ocorrer se o time-based stop depende de trade.open_date_utc e current_time, mas um deles estiver em timezone diferente ou sem valor correto.
	•	Verificar:
	•	No custom_stoploss, o trecho:

dur_hours = (current_time - trade.open_date_utc).total_seconds() / 3600.0
if dur_hours >= self.max_trade_duration_in_hours.value:
    return -0.0001

	•	Se essa conta está de fato sendo executada. Pode-se inserir um debug print para checar os valores de current_time, trade.open_date_utc e dur_hours.

4. Uso de force_exit ou liquidation
	•	Sintoma: No relatório, às vezes vemos exit_reason = force_exit ou liquidation, indicando que a time-based stop não foi propriamente atingida, mas sim algum “evento” externo (p. ex. a simulação de liquidação em Futures).
	•	Explicação: Se em um backtest de Futures a perda ultrapassar um certo nível, o Freqtrade pode simular “liquidation” em vez de acionar seu stop. Se esse valor for configurado muito alto, isso permitiria trades chegarem a -80% ou mais.
	•	Como checar:
	•	Ver se no config.json há algo como "unrealized_profit_threshold": ... ou configuração de liquidação simulada.
	•	Ou se o Freqtrade detecta a perda como irreversível e “liquida” o trade.

5. stoploss e Distâncias de ATR
	•	Sintoma: Se o stoploss (ou seja, o SL gerado pelo calc_takeprofit_stoploss_v4) for muito distante e nunca for atingido, um trade pode ficar “travado” e só sair quando finalmente encerrar por tempo ou force_exit.
	•	Confirmar:
	•	Se max_stop_factor não está gigantesco (p. ex. 5.0 ou 6.0), pois isso pode permitir SL bem distante do preço.
	•	Se min_stop_factor não está muito baixo, o que poderia gerar stop pequeno. Mas no seu caso, pode estar criando stop muito longe devido a ATR grande.

6. force_exit ao final do timerange
	•	Sintoma: No final do período de backtest (--timerange), o Freqtrade encerra automaticamente qualquer trade “aberto” como force_exit.
	•	Efeito: Se o período total for de ~3 anos, e o trade iniciou logo cedo no backtest mas nunca atingiu o SL/TP, pode ficar aberto até o fim do timerange, resultando em ~600+ dias.
	•	Solução: Se deseja evitar esse cenário:
	•	Ajustar max_trade_duration_in_hours bem menor do que a duração do backtest.
	•	Confirmar que a “time-based stop” seja usada de fato.

7. Resumo de como solucionar/troubleshoot
	1.	Checar use_custom_stoploss:
	•	Confirme se está true no config.json. Em versões atuais do Freqtrade, bastaria não desativá-lo. Se use_custom_stoploss estiver false, o custom_stoploss não é chamado.
	2.	Baixar max_trade_duration_in_hours:
	•	Se 72h é o alvo, qualquer trade >72h deve ser encerrado. Se ainda há trades de 56 dias, algo não está sendo respeitado.
	3.	Incluir debug prints no custom_stoploss:
	•	Especialmente após calcular dur_hours. Exemplo:

if debug:
    print(f"Trade {trade.id} dur_hours={dur_hours:.2f}, max={self.max_trade_duration_in_hours.value}")

	•	Assim, ao rodar o backtest, você verifica se esse trecho está de fato sendo acionado.

	4.	Ver se há menção a “liquidation”:
	•	Um trade que chega a -80% é sinal de liquidação simulada ou stop não sendo respeitado.
	5.	Confirmar que max_stop_factor não está imenso:
	•	Se max_stop_factor = 5.0, com um ATR grande, um STOP pode ficar absurdamente longe, e o preço eventualmente vai “derreter” muito antes de bater no STOP.
	•	Ajustar para 2.0 ou 3.0, conforme o mercado.
	6.	Ver config de Futures:
	•	Em Futures, a alavancagem e “liquidação” podem sim gerar trades com perda muito maior do que -10% ou -20%. Avalie como o backtester está simulando essa liquidação.

Conclusão

O fato de ver Winners ~4 dias vs. Losers ~56 dias normalmente aponta que alguns trades perdedores não dispararam stop (ou time-based stop) como esperado. As razões mais prováveis são:
	•	O time-based stop não está sendo acionado (possivelmente use_custom_stoploss desabilitado, ou dur_hours não sendo calculado corretamente).
	•	O stoploss foi colocado longe devido a ATR alta + max_stop_factor generosa, e o preço demorou a chegar lá, sendo encerrado apenas no final do timerange ou via “liquidação”.
	•	Houve force_exit no final do backtest, gerando durações artificialmente grandes.

Verifique cada um dos pontos acima no seu ambiente e, ao corrigi-los, você deverá ver durações de loss mais adequadas (e sem trades de 50+ dias).
